node/network/gossip-support/src/lib.rs

```diff
@@ -157,6 +168,80 @@ pub async fn connect_to_authorities(
 	failed_rx
 }
 
+/// We partition the list of all sorted `authorities` into sqrt(len) groups of sqrt(len) size
+/// and form a matrix where each validator is connected to all validators in its row and column.
+/// This is similar to [web3] research proposed topology, except for the groups are not parachain
+/// groups (because not all validators are parachain validators and the group size is small),
+/// but formed randomly via BABE randomness from two epochs ago.
+/// This limits the amount of gossip peers to 2 * sqrt(len) and ensures the diameter of 2.
+///
+/// [web3]: https://research.web3.foundation/en/latest/polkadot/networking/3-avail-valid.html#topology
+async fn update_gossip_topology(
+	ctx: &mut impl SubsystemContext,
+	our_index: usize,
+	authorities: Vec<AuthorityDiscoveryId>,
+	relay_parent: Hash,
+) -> Result<(), util::Error> {
+	// retrieve BABE randomness
+	let random_seed = {
+		let (tx, rx) = oneshot::channel();
+
+		ctx.send_message(RuntimeApiMessage::Request(
+			relay_parent,
+			RuntimeApiRequest::CurrentBabeEpoch(tx),
+		).into()).await;
+
+		let randomness = rx.await??.randomness;
+		let mut subject = [0u8; 40];
+		subject[..8].copy_from_slice(b"gossipsu");
+		subject[8..].copy_from_slice(&randomness);
+		sp_core::blake2_256(&subject)
+	};
+
+	// shuffle the indices
+	let mut rng: ChaCha20Rng = SeedableRng::from_seed(random_seed);
+	let len = authorities.len();
+	let mut indices: Vec<usize> = (0..len).collect();
+	indices.shuffle(&mut rng);
+	let mut reverse_indices = vec![0usize; len];
+	for (i, idx) in indices.iter().enumerate() {
+		reverse_indices[*idx] = i;
+	}
+
+	let neighbors = matrix_neighbors(reverse_indices[our_index], len);
+	let our_neighbors = neighbors.map(|i| authorities[indices[i]].clone()).collect();
+
+	ctx.send_message(AllMessages::NetworkBridge(
+		NetworkBridgeMessage::NewGossipTopology {
+			our_neighbors,
+		}
+	)).await;
+
+	Ok(())
+}
+
+/// Compute our row and column neighbors in a matrix
+fn matrix_neighbors(our_index: usize, len: usize) -> impl Iterator<Item=usize> {
+	assert!(our_index < len, "our_index is computed using `enumerate`; qed");
+
+	// e.g. for size 11 the matrix would be
+	//
+	// 0  1  2
+	// 3  4  5
+	// 6  7  8
+	// 9 10
+	//
+	// and for index 10, the neighbors would be 1, 4, 7, 9
+
+	let sqrt = (len as f64).sqrt() as usize;
+	let our_row = our_index / sqrt;
+	let our_column = our_index % sqrt;
+	let row_neighbors = our_row * sqrt..std::cmp::min(our_row * sqrt + sqrt, len);
+	let column_neighbors = (our_column..len).step_by(sqrt);
```

Comment:
```suggestion
	let pos = |r, c| { let p = (r * sqrt) + c; if p != our_pos && p < len {Some(p) } else { None };
	let row_neighbors = (0..sqrt).filter_map(|c| pos(our_row, c));
	let column_neighbors = (0..sqrt).filter_map(|r| pos(r, our_column));
```

Reply:
nit: i find it slightly easier to read this way

Reply:
it doesn't compile and is not what we want, because it's actually not a matrix (number of rows might be higher and sqrt)